### PR TITLE
Record Data State

### PR DESCRIPTION
**Pull request by igorT**

### CODE REVIEW

**Comment by igorT** [comment]
Has a few failing tests I'm addressing

**Comment by igorT** [comment]
All tests should be passing now

File: packages/store/addon/-private/system/store.js [hunk 1]
```diff
@@ -2126,15 +2133,26 @@ const Store = Service.extend({
...
 
-      if (internalModel.currentState.stateName === 'root.deleted.saved') {
-        resolver.resolve();
-        continue;
-      } else if (internalModel.isNew()) {
-        operation = 'createRecord';
-      } else if (internalModel.isDeleted()) {
-        operation = 'deleteRecord';
+      if (!ENABLE_RECORD_DATA_STATE) {
+        if (internalModel.currentState.stateName === 'root.deleted.saved') {
```
Comment (runspired) [inline packages/store/addon/-private/system/store.js]: where does the handling of this state move?

File: packages/store/addon/-private/system/model/internal-model.ts [hunk 2]
```diff
@@ -202,15 +204,45 @@ export default class InternalModel {
     // `lastObject` have changed.  When this happens we don't want those
     // models to rematerialize their records.
 
+    // eager checks to avoid instantiating record data if we are empty or loading
+    if (this.isEmpty()) {
+      return true;
+    }
+
+    if (this.isLoading()) {
```
Comment (runspired) [inline packages/store/addon/-private/system/model/internal-model.ts]: this seems to be a new addition to our logic?

File: packages/store/addon/-private/system/model/internal-model.ts [hunk 3]
```diff
@@ -202,15 +204,45 @@ export default class InternalModel {
...
+    let isRecordFullyDeleted;
+    if (ENABLE_RECORD_DATA_STATE) {
+      isRecordFullyDeleted = this._isRecordFullyDeleted();
+    } else {
+      isRecordFullyDeleted = this.currentState.stateName === 'root.deleted.saved';
+    }
     return (
-      this._isDematerializing ||
       this.hasScheduledDestroy() ||
+      this._isDematerializing ||
```
Comment (runspired) [inline packages/store/addon/-private/system/model/internal-model.ts]: why the ordering switch?

File: packages/-ember-data/tests/integration/record-data/record-data-state-test.ts [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+
+  isNew() {
+    return false;
+  }
+
+  isDeleted() {
+    return false;
+  }
+
+  addToHasMany(key: string, recordDatas: this[], idx?: number) { }
```
Comment (mike-north) [inline packages/-ember-data/tests/integration/record-data/record-data-state-test.ts]: replace `this` with `RecordData`

File: packages/-ember-data/tests/integration/record-data/record-data-state-test.ts [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+    owner.register('service:store', CustomStore);
+  });
+
+  test("Record Data state saving", async function (assert) {
+    assert.expect(3);
+    let isDeleted, isNew, isDeletionCommitted;
+    let calledDelete = false;
+    let calledUpdate = false;
+    let calledCreate = false;
+    let storeWrapper;
```
Comment (mike-north) [inline packages/-ember-data/tests/integration/record-data/record-data-state-test.ts]: with no initializer, will be an `any` unless you provide a type

File: packages/store/addon/-private/system/model/internal-model.ts (hunk 2 above)
Comment (igorT) [reply packages/store/addon/-private/system/model/internal-model.ts]: I flagged it, but the addition is due to `_isRecordFullyDeleted` accessing recordData, which if you are `isLoading` means that you will eagerly instantiate the RD, even though we know none of the other properties should be true. Great call on flagging though, as it turns out right now all of the tests pass both when returning false and when returning true. I will make sure to add a test on master that asserts the correct behavior

File: packages/store/addon/-private/system/model/internal-model.ts (hunk 3 above)
Comment (igorT) [reply packages/store/addon/-private/system/model/internal-model.ts]: debugging leftover, reverting to original

**Comment by igorT** [comment]
@runspired great catch on model fragments. The test fails because they are using our record data, creating a fragment using a path that makes it think it's been pushed, and then rewiring the state machine to show that the fragment is `isNew`. So it should be pretty straightforward to fix and make their code nicer when this lands

File: packages/store/addon/-private/system/store.js (hunk 1 above)
Comment (igorT) [reply packages/store/addon/-private/system/store.js]: https://github.com/emberjs/data/pull/6204/files#diff-1080751ea355f7ab0c457d1785c50695R2103

**Pull request merged**